include/uk/plat/lcpu.h

```diff
@@ -96,100 +96,129 @@ void ukplat_lcpu_halt_to(__nsec until);
-	 *   function execution
+	 *   to the state are applied after the RUN IRQ handler returns
+	 * @param fn pointer to this structure
 	 */
-	void (*fn)(struct __regs *regs, struct ukplat_lcpu_func *fn);
+	void (*fn)(struct __regs *regs, const struct ukplat_lcpu_func *fn);
 
 	/* Optional user-supplied pointer. */
 	void *user;
 };
 
 /**
- * Returns the ID of the current logical CPU
+ * Returns the index of the current logical CPU
  */
-__lcpuid ukplat_lcpu_id(void);
+__lcpuidx ukplat_lcpu_idx(void);
 
 /**
- * Returns the number of logical CPUs present on the system
+ * Returns the number of logical CPUs present in the system
  */
-__lcpuid ukplat_lcpu_count(void);
+__u32 ukplat_lcpu_count(void);
 
 /**
  * Starts multiple logical CPUs and assigns them the given stacks. The logical
  * CPUs execute the entry functions if supplied or enter a low-power wait state
- * otherwise.
+ * otherwise. CPUs that have already been started are ignored.
  *
- * @param lcpuid array with the IDs of the logical CPUs that are to be started
- * @param sp array of stack pointers, one for each specified logical CPU
- * @param entry optional array of entry function pointers, one for each
- *   specified logical CPU. Provided functions must not return. If the
- *   parameter or individual elements are NULL the respective logical CPUs
- *   enter a low-power wait state after startup
- * @param num number of logical CPU IDs given
+ * @param lcpuidx array with the indices of the logical CPUs that are to be
+ *   started. CPUs are started in the order specified in the array. Can be NULL
+ *   to include all logical CPUs except the one executing the function, in which
+ *   case CPUs are started in sequential order according to their CPU index.
+ * @param[inout] num if lcpuidx is not NULL, provides [IN] the number of
+ *   elements in lcpuidx, and [OUT] the number of successfully started CPUs in
+ *   sequential order of lcpuidx. If the call succeeds, input and output values
+ *   are equal. Must be NULL if lcpuidx is NULL.
+ * @param sp array of stack pointers, one for each logical CPU to start. If
+ *   lcpuidx is NULL, must be ukplat_lcpu_count() - 1 stacks
+ * @param entry array of entry function pointers, one for logical CPU to start.
+ *   Can be NULL, otherwise if lcpuidx is NULL, must contain
+ *   ukplat_lcpu_count() - 1 function pointers. Provided functions must not
+ *   return. If the parameter or individual function pointers are NULL the
+ *   respective logical CPUs enter a low-power wait state after startup
+ * @param flags (architecture-dependent) flags that specify how to start the
+ *   CPUs (see UKPLAT_LCPU_SFLG_* flags if available)
  *
  * @return 0 on success, an errno-type error value otherwise
  */
-int ukplat_lcpu_start(__lcpuid lcpuid[], void *sp[],
-		      ukplat_lcpu_entry_t entry[], unsigned int num);
+int ukplat_lcpu_start(const __lcpuidx lcpuidx[], unsigned int *num, void *sp[],
```

Comment:
Actually, I decided against it, because I think it makes sense to give full control of what to do with the CPUs. As we talked on Discord, this would imply that we have to also make `lcpu_init` public. But you can avoid all of this with your suggestion to always start in the `default` entry for secondary CPUs, do the `lcpu_init` there and then jump out to the supplied entry function if there is one.

If you need a specifically prepared stack you can still prepare it in the supplied entry function and then do the actual jump to the target function. So the user would do:
```C
int main(int argc, char* argv[]) {
   Run me on a diffierent core
}

void __noreturn myentry(void *arg) {
   PREPARE stack with info from arg
   jump to main
}

void *stackp = malloc(STACK_SIZE);
ukplat_lcpu_entry_func fn = {
  .fn = myentry;
 .arg = Some info to prepare stack, or actually pointer to prepared stack
} 
ukplat_lcpu_start(othercpu, 1, stackp, fn, 0);
```
But in fact the call chain is like this:
`ASM trampoline` -> `entry_default` -> `myentry`

Reply:
> Actually, I decided against it, because I think it makes sense to give full control of what to do with the CPUs.

Okay for this PR, but eventually we should really provide a clear definition of that. To me it sounds paradoxical to start the cores from the application and at the same time have no access to drivers or other low-level facilities. Or we stick to the current scheme for consistency, and expect that any application that needs to start the cores has to manually add the internal header into its `CINCLUDES-y` into its Makefile.uk. This is what we do for other private interfaces.

Reply:
Ok, so you would still make this private? I think this whole flexibility to defined per-cpu custom entry functions looses a bit of power if you keep it in the platform. May @skuenzer could add a third opinion on this?

Reply:
> Ok, so you would still make this private?

I wouldn't block the PR over this one, it's not such a big deal 🙂 My point is more towards agreeing on the necessity to resolve this and other core architectural topics before too much debt is accumulated.

Reply:
I had a short discussion with Simon about this. He would also prefer this to be public. But based on this discussion I will also keep the entry function typedef as we have it at the moment.

But agree, we have to discuss this more as part of our platform restructure so that we have better guidelines for what APIs to have at what level.